### PR TITLE
[foreach_map] Add foreach_map Adam impl to compiled optimizer tests

### PR DESCRIPTION
**Pull request by mlazos**
Adds a foreach_map backed Adam to compiled optimizer tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: test/inductor/test_compiled_optimizers.py [hunk 1]
```diff
@@ -63,6 +64,82 @@
...
+    with torch.no_grad():
+        grads = [param.grad for param in params]
+        # update step
+        updated_steps = foreach_map(lambda x: x + 1, (steps,))
+        torch._foreach_copy_(steps, updated_steps)
+
+        if weight_decay != 0:
+            foreach_map(torch.add, (grads,), alpha=weight_decay)
+
+        # HOPS cannot have multiple outputs at the moment
```
Comment (Chillee) [inline test/inductor/test_compiled_optimizers.py:117]: Would be good to resolve this lol

**Review by Chillee — APPROVED** [review]

File: test/inductor/test_compiled_optimizers.py [hunk 2]
```diff
@@ -63,6 +64,82 @@
...
+        # need to call foreach_map once for each output
+        exp_avgs_updated = foreach_map(torch.lerp, (exp_avgs, grads, 1 - beta1))
```
Comment (eellison) [inline test/inductor/test_compiled_optimizers.py:119]: cc @zou3519

**Review by eellison — APPROVED** [review]
niceeee

**Pull request closed without merging**